guidelines/enhancement_template.md

```diff
@@ -311,15 +316,31 @@ Upgrade expectations:
   pods must continue to work correctly even while the cluster remains
   in this partially upgraded state for some time.
 
-Downgrade expectations:
-- If an `N->N+1` upgrade fails mid-way through, or if the `N+1` cluster is
-  misbehaving, it should be possible for the user to rollback to `N`. It is
-  acceptable to require some documented manual steps in order to fully restore
-  the downgraded cluster to its previous state. Examples of acceptable steps
-  include:
-  - Deleting any CVO-managed resources added by the new version. The
-    CVO does not currently delete resources that no longer exist in
-    the target version.
+#### Downgrade strategy
+
+If an `N->N+1` upgrade fails mid-way through, or if the `N+1` cluster is
+misbehaving, users may wish to rollback to `N`.  We [do not support
```

Comment:
MicroShift deployments using ostree images may reboot back into older images if the GreenBoot checks fail. MicroShift itself will need to deal with that to some extent, but any component that stores state outside of etcd (for example, on a PV) also needs to be able to handle this rollback scenario. Granted that's a very small set of use cases for most of what makes it into this repo, but in the context of these design documents we cannot make the blanket statement that downgrades are never supported under any circumstances.

Reply:
> MicroShift itself will need to deal with that to some extent, but any component that stores state outside of etcd...

And etcd too, right?  They've had a few etcd minor bumps that triggered on-disk format changes, where the outgoing etcd version was not able to read that on-disk format written by the incoming etcd version.

But the enhancement-author ask is still going to be "try and make it clean, but if not, at least explain yourself".  Can I soften the no-support statement with something like:

> OCP [does not support rollbacks or downgrades of any kind][ocp-no-downgrades], but if structuring to
allow rollbacks is possible, that is a nice internal safety valve.  And MicroShift [does support some rollbacks][microshift-rollback].

?

[ocp-no-downgrades]: https://github.com/openshift/openshift-docs/blob/7f87267bc69d65abd96e6b783100195c6b78549f/updating/updating-troubleshooting.adoc
[microshift-rollback]: https://github.com/openshift/microshift/blame/1eb5c442d2661086f32c3768b2e6ac8c6e2cb6fa/README.md#L9

Reply:
The new language is better. 

I'm not excited to support the statements on line 327-328 that it's OK to break clusters or lock up. If we can just say "automated downgrades are not a requirement, but please provide insight about why manual downgrades may be hard or impossible for this change", without the list?

What sort of help does the customer receive when they contact support, as directed in the docs you link there? Do we walk them through recovering a previous state from a backup, for example?

Reply:
> Do we walk them through recovering a previous state from a backup, for example?

Possibly, although [our restore-from-backup docs][1] say:

> Restoring to a previous cluster state is a destructive and destablizing action to take on a running cluster. This should only be used as a last resort.
>
> If you are able to retrieve data using the Kubernetes API server, then etcd is available and you should not restore using an etcd backup.

And support can point folks [at][2]:

> Rolling your cluster back to a previous version is not supported.

Except on MicroShift, where... :shrug: you probably understand how MicroShift support would handle these better than me.  And "Red Hat support helps you install a fresh cluster and migrate your workloads over to it" is always on the table as a safety-valve option.

And I'm not clear on if/how MicroShift is planning to audit enhancements for rollback exposure.  Are they expecting the full set of enhancement approvers to be looking out for them (which would require the full set of approvers to understand what MicroShift exposure looks like)?  Or are MicroShift-aware approvers expected to be called in to review any enhancement that doesn't claim clean, beautiful automatic downgrades that automatically removing any `N+1`-specific resources?  Or?

[1]: https://docs.openshift.com/container-platform/4.11/backup_and_restore/control_plane_backup_and_restore/disaster_recovery/scenario-2-restoring-cluster-state.html
[2]: https://github.com/openshift/openshift-docs/blob/7f87267bc69d65abd96e6b783100195c6b78549f/updating/updating-troubleshooting.adoc

Reply:
> > Restoring to a previous cluster state is a destructive and destablizing action to take on a running cluster. This should only be used as a last resort.
> > If you are able to retrieve data using the Kubernetes API server, then etcd is available and you should not restore using an etcd backup.
> 
> And support can point folks [at](https://github.com/openshift/openshift-docs/blob/7f87267bc69d65abd96e6b783100195c6b78549f/updating/updating-troubleshooting.adoc):
> 
> > Rolling your cluster back to a previous version is not supported.
> 
> Except on MicroShift, where... 🤷 you probably understand how MicroShift support would handle these better than me. And "Red Hat support helps you install a fresh cluster and migrate your workloads over to it" is always on the table as a safety-valve option.

Right, the support process will be different for MicroShift, but will also include "start over" as an option.

> And I'm not clear on if/how MicroShift is planning to audit enhancements for rollback exposure. Are they expecting the full set of enhancement approvers to be looking out for them (which would require the full set of approvers to understand what MicroShift exposure looks like)? Or are MicroShift-aware approvers expected to be called in to review any enhancement that doesn't claim clean, beautiful automatic downgrades that automatically removing any `N+1`-specific resources? Or?

MicroShift doesn't rely on most of the operators in OCP, so it shouldn't need to be everyone. The etcd, control plane, LVMS, and OVN-K teams all have representatives working closely with the MicroShift team and should provide good coverage for rollback concerns. Owners of individual add-on operators will also be aware of whether their operator is supported on MicroShift.